WorkflowSwiftUI/Sources/ObservableScreen.swift

```diff
@@ -0,0 +1,208 @@
+///
+/// Screens conforming to this protocol will render SwiftUI views that observe fine-grained changes
+/// to the underlying model, and selectively invalidate in response to changes to properties that
+/// are accessed by the view.
+///
+/// Invalidations happen when the observed state is mutated, during actions or the
+/// `workflowDidChange` method. When this screen is rendered, a new model is injected into the
+/// store. Any invalidated views will then be updated with the new model by SwiftUI during its own
+/// rendering cycle.
+///
+/// To use this protocol with a workflow, your workflow should render a type that conforms to
+/// ``ObservableModel``, and then mapped to a screen implementation that uses that concrete model
+/// type. See ``ObservableModel`` for options on how to render one easily.
+public protocol ObservableScreen: Screen {
+    /// The type of the root view rendered by this screen.
+    associatedtype Content: View
+    /// The type of the model that this screen observes.
+    associatedtype Model: ObservableModel
+
+    /// The sizing options for the screen.
+    var sizingOptions: SwiftUIScreenSizingOptions { get }
+    /// The model that this screen observes.
+    var model: Model { get }
+
+    /// Constructs the root view for this screen. This is only called once to initialize the view.
+    /// After the initial construction, the view will be updated by injecting new values into the
+    /// store.
+    @ViewBuilder
+    static func makeView(store: Store<Model>) -> Content
+}
+
+public extension ObservableScreen {
+    var sizingOptions: SwiftUIScreenSizingOptions { [] }
+}
+
+public extension ObservableScreen {
+    func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
+        ViewControllerDescription(
+            type: ModeledHostingController<Model, Content>.self,
+            environment: environment,
+            build: {
+                let (store, setModel) = Store.make(model: model)
+                return ModeledHostingController(
+                    setModel: setModel,
+                    viewEnvironment: environment,
+                    rootView: Self.makeView(store: store),
+                    sizingOptions: sizingOptions
+                )
+            },
+            update: { hostingController in
+                hostingController.setModel(model)
+                hostingController.setViewEnvironment(environment)
+            }
+        )
+    }
+}
+
+public struct SwiftUIScreenSizingOptions: OptionSet {
+    public let rawValue: Int
+
+    public init(rawValue: Int) {
+        self.rawValue = rawValue
+    }
+
+    public static let preferredContentSize: SwiftUIScreenSizingOptions = .init(rawValue: 1 << 0)
+}
+
+private struct ViewEnvironmentModifier: ViewModifier {
+    @ObservedObject var holder: ViewEnvironmentHolder
+
+    func body(content: Content) -> some View {
+        content
+            .environment(\.viewEnvironment, holder.viewEnvironment)
```

Comment:
Confirming it's OK to stamp the entire view env modifier over the existing one? No way anything in the middle can change it?

Reply:
Makes sense!